File: tools/notifications.py
from __future__ import annotations

import urllib.request
import urllib.error
import json
from config import settings


def _post(payload) -> None:
    """POST a payload to the n8n webhook, silently no-op'ing if it's unset.

    Two payload shapes are supported:
    - **dict** — serialised verbatim. Used by the rules-engine pivot's
      structured event types (#196), which encode `event_type` plus
      event-specific fields so downstream automations can route on shape.
    - **str** — wrapped as ``{"message": <str>}`` before serialising.
      Preserved for ``notify_error`` and ``notify_panic`` which both post
      free-form Discord-style strings.
    """
    if not settings.N8N_WEBHOOK_URL:
        return
    if isinstance(payload, dict):
        body = json.dumps(payload).encode()
    else:
        body = json.dumps({"message": payload}).encode()
    req = urllib.request.Request(
        settings.N8N_WEBHOOK_URL,
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    try:
        urllib.request.urlopen(req, timeout=5)
    except Exception as e:
        print(f"[notifications] failed to send: {e}")


def notify_error(context: str, error: str) -> None:
    if len(error) <= 500:
        snippet = error
    else:
        snippet = error[:240] + "\n...\n" + error[-240:]
    _post(f"⚠️ **Bot Error — {context}**\n```{snippet}```")


def notify_panic(action: str, results: list, dry_run: bool = False) -> None:
    """Post a Discord alert summarising a `main.py panic` invocation.

    `action` is the human-readable headline ("cancel-orders", "liquidate",
    "pause", or a composite). `results` is a list of dicts (typically the
    return values of `cancel_all_orders` / liquidate); each row is rendered
    as a single line. `dry_run=True` flags the alert as a preview — used by
    `--liquidate` without `--confirm`.
    """
    suffix = " (DRY RUN)" if dry_run else ""
    lines = [f"🛑 **PANIC — {action}**{suffix}"]
    if not results:
        lines.append("📭 nothing to do")
    else:
        lines.append(f"📊 {len(results)} item{'s' if len(results) != 1 else ''}:")
        for r in results:
            symbol = r.get("symbol") or r.get("ticker") or ""
            order_id = r.get("order_id") or r.get("id") or ""
            status = r.get("status")
            descriptor = " ".join(p for p in [symbol, str(order_id) if order_id else "", f"[{status}]" if status is not None else ""] if p)
            lines.append(f"• {descriptor or r}")
    _post("\n".join(lines))


# ---------------------------------------------------------------------------
# Rules-engine pivot (#196): structured-payload event types.
#
# Each helper posts a JSON dict ({"event_type": "...", ...fields}) so the
# n8n flow downstream can route on shape rather than parsing free-form
# Discord prose. Keyword-only signatures keep call sites self-documenting.
# ---------------------------------------------------------------------------


def notify_regime_flip(
    *,
    target_state: str,
    spy_close: float,
    spy_sma200: float,
    ticker: str,
    fill_price: float,
    qty: int,
    account_value: float,
    dry_run: bool = False,
) -> None:
    """Emitted whenever the SPY/SMA200 regime filter flips state and we trade
    on it (LONG entry or CASH exit). Includes the SPY snapshot that drove the
    decision plus the resulting fill so downstream audit can reconcile.

    When ``dry_run=True``, the payload includes ``dry_run: true`` and the
    ``title`` field is prefixed with ``[DRY-RUN]`` so n8n / Discord templates
    can branch on it. Used by ``daily_check.py --dry-run`` to publish a
    hypothetical flip alert during the soak window without placing an order.
    """
    title_prefix = "[DRY-RUN] " if dry_run else ""
    _post({
        "event_type": "regime_flip",
        "title": f"{title_prefix}regime_flip {target_state}",
        "target_state": target_state,
        "spy_close": spy_close,
        "spy_sma200": spy_sma200,
        "ticker": ticker,
        "fill_price": fill_price,
        "qty": qty,
        "account_value": account_value,
        "dry_run": dry_run,
    })


def notify_kill_switch_fired(
    *,
    ticker: str,
    drawdown_pct: float,
    ref_high: float,
    last_price: float,
    qty: int,
    fill_price: float,
) -> None:
    """Emitted when the per-position drawdown kill switch closes a holding.
    `drawdown_pct` is signed (negative for a loss); `ref_high` is the high-
    water mark used as the drawdown reference.
    """
    _post({
        "event_type": "kill_switch_fired",
        "ticker": ticker,
        "drawdown_pct": drawdown_pct,
        "ref_high": ref_high,
        "last_price": last_price,
        "qty": qty,
        "fill_price": fill_price,
    })


def notify_trade_failed(
    *,
    symbol: str,
    side: str,
    qty: int,
    reason: str,
) -> None:
    """Emitted when the broker rejects an order. `reason` is the broker's
    (or our pre-submit gate's) machine-readable code, e.g.
    ``"insufficient_buying_power"``.
    """
    _post({
        "event_type": "trade_failed",
        "symbol": symbol,
        "side": side,
        "qty": qty,
        "reason": reason,
    })


def notify_tws_disconnected(
    *,
    host: str,
    port: int,
    attempts: int,
    error_msg: str,
) -> None:
    """Emitted when the IBKR TWS/Gateway connection cannot be re-established
    after `attempts` retries. `error_msg` is the last error string from the
    underlying client.
    """
    _post({
        "event_type": "tws_disconnected",
        "host": host,
        "port": port,
        "attempts": attempts,
        "error_msg": error_msg,
    })


def notify_state_desync(
    *,
    db_state: str,
    broker_state: str,
    symbol: str,
    action_taken: str,
) -> None:
    """Emitted when the persisted `regime_state` (DB) disagrees with broker
    truth on reconciliation. `action_taken` describes the corrective step
    the bot performed (e.g. ``"DB updated to CASH"``).
    """
    _post({
        "event_type": "state_desync",
        "db_state": db_state,
        "broker_state": broker_state,
        "symbol": symbol,
        "action_taken": action_taken,
    })

File: supabase/functions/_shared/notifications.ts
// Direct-to-Discord webhook poster (#362 — no n8n middleman). Every event
// carries structured `event_type` + event-specific fields (kept for any
// future JSON-consuming forwarder), plus a human-readable `message`. When
// `message` is present, postEvent() also derives a Discord-native `content`
// field (Discord's incoming-webhook API renders `content` directly, and
// ignores unrecognised extra fields), codepoint-safe-truncated to 2,000
// characters — Discord's hard content limit — so a long message still posts
// instead of 400ing the whole webhook call. The full untruncated text always
// survives in `message`. NEVER throws — a notification outage must not crash
// the bot. An unset secret, a fetch rejection, or a non-2xx response are all
// logged via console.warn (visible in Supabase function logs, #366) — never
// the webhook URL or the full payload — so an outage is silent to the caller
// but visible in ops.
//
// #397: postEvent() is notify()'s former body, now returning a NotifyStatus
// instead of discarding it, so outbox.ts's notifyDurable can enqueue a
// "failed" post for retry. notify() stays a thin `Promise<void>` wrapper —
// its signature and observable behavior are UNCHANGED (see notifications.test.ts,
// every pre-#397 case still passes unmodified) because widening notify()'s own
// return type would force edits to the DailyCheckDeps/KillSwitchDeps
// notifications interfaces in both logic.ts files, which must stay untouched.
import { getNotifyWebhookUrl } from "./config.ts";

const DISCORD_CONTENT_MAX_CODEPOINTS = 2000;
const WARN_BODY_SNIPPET_MAX_CODEPOINTS = 200;

function truncateCodepoints(s: string, max: number): string {
  return [...s].slice(0, max).join("");
}

export type NotifyStatus = "sent" | "failed" | "skipped_unset";

export async function postEvent(event: Record<string, unknown>): Promise<NotifyStatus> {
  const url = getNotifyWebhookUrl();
  const eventType = String(event.event_type ?? "unknown");
  if (url === "") {
    console.warn(`notify: skipped (NOTIFY_WEBHOOK_URL unset), event_type=${eventType}`);
    return "skipped_unset";
  }
  const message = event.message;
  const body = typeof message === "string" && message !== ""
    ? { ...event, content: truncateCodepoints(message, DISCORD_CONTENT_MAX_CODEPOINTS) }
    : event;
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      const text = await res.text();
      // Redact the FULL untruncated text first, then truncate — the reverse
      // order would leave an un-redacted prefix of the URL (including part of
      // any secret token) in the snippet whenever the URL straddles the cut.
      const snippet = truncateCodepoints(
        text.replaceAll(url, "[webhook-url]"),
        WARN_BODY_SNIPPET_MAX_CODEPOINTS,
      );
      console.warn(
        `notify: webhook responded ${res.status}: ${snippet}, event_type=${eventType}`,
      );
      return "failed";
    }
    return "sent";
  } catch (e) {
    const name = e instanceof Error ? e.name : "Error";
    const msg = String(e instanceof Error ? e.message : e).replaceAll(url, "[webhook-url]");
    console.warn(`notify: webhook POST failed: ${name}: ${msg}, event_type=${eventType}`);
    return "failed";
  }
}

export async function notify(event: Record<string, unknown>): Promise<void> {
  await postEvent(event);
}

export function regimeFlipEvent(p: {
  targetState: "LONG" | "CASH";
  spyClose: number;
  spySma200: number;
  ticker: string;
  fillPrice: number;
  qty: number;
  accountValue: number;
  dryRun?: boolean;
}): Record<string, unknown> {
  const titlePrefix = p.dryRun ? "[DRY-RUN] " : "";
  const message = `${titlePrefix}Regime flip -> ${p.targetState}: ${p.ticker} qty=${p.qty} ` +
    `@ $${p.fillPrice.toFixed(2)} (SPY $${p.spyClose.toFixed(2)} vs 200-DMA $${
      p.spySma200.toFixed(2)
    })`;
  return {
    event_type: "regime_flip",
    title: `${titlePrefix}regime_flip ${p.targetState}`,
    message,
    target_state: p.targetState,
    spy_close: p.spyClose,
    spy_sma200: p.spySma200,
    ticker: p.ticker,
    fill_price: p.fillPrice,
    qty: p.qty,
    account_value: p.accountValue,
    dry_run: p.dryRun ?? false,
  };
}

export function notifyRegimeFlip(p: Parameters<typeof regimeFlipEvent>[0]): Promise<void> {
  return notify(regimeFlipEvent(p));
}

// #474 T4 (spec §8.1): side-aware -- refPrice is side-neutral (the rolling
// high for a LONG fire, the rolling low for a SHORT fire) so the same event
// shape covers the kill-switch's short mirror. side names which close verb
// fired (SELL closes a long, BUY covers a short).
export function killSwitchFiredEvent(p: {
  ticker: string;
  side: "LONG" | "SHORT";
  drawdownPct: number;
  refPrice: number;
  lastPrice: number;
  qty: number;
  fillPrice: number;
}): Record<string, unknown> {
  const closeVerb = p.side === "LONG" ? "SELL" : "BUY";
  const refLabel = p.side === "LONG" ? "ref high" : "ref low";
  const message =
    `Kill switch fired on ${p.ticker} (${p.side}): drawdown ${(p.drawdownPct * 100).toFixed(1)}% ` +
    `(${refLabel} $${p.refPrice.toFixed(2)}, last $${p.lastPrice.toFixed(2)}), ` +
    `${closeVerb} qty=${p.qty} @ $${p.fillPrice.toFixed(2)}`;
  return {
    event_type: "kill_switch_fired",
    message,
    ticker: p.ticker,
    side: p.side,
    drawdown_pct: p.drawdownPct,
    ref_price: p.refPrice,
    last_price: p.lastPrice,
    qty: p.qty,
    fill_price: p.fillPrice,
  };
}

export function notifyKillSwitchFired(
  p: Parameters<typeof killSwitchFiredEvent>[0],
): Promise<void> {
  return notify(killSwitchFiredEvent(p));
}

// #514: the -15% equity floor's halt event (EQUITY_FLOOR_PCT in
// hourly-check/logic.ts). floorPct is fixed at 0.15 here rather than taken
// from the caller -- the floor is a spec-frozen constant (§11), not a config
// setting, so there is nothing for a caller to legitimately vary. position at
// pause time is the caller's job to resolve (a fresh, guarded broker read);
// this function only renders whatever it is handed, including "unknown".
export function equityFloorFiredEvent(p: {
  ticker: string;
  equity: number;
  baseline: number;
  drawdownPct: number;
  /** null = position read failed or was never attempted -- rendered as unknown/treat-as-open. */
  positionQty: number | null;
  /** Set only when positionQty is null because the read itself failed. */
  positionError?: string;
}): Record<string, unknown> {
  const head =
    `EQUITY FLOOR FIRED on ${p.ticker}: equity $${p.equity} vs baseline $${p.baseline} ` +
    `(-${(p.drawdownPct * 100).toFixed(1)}%). bot_config.paused=true.`;
  let tail: string;
  if (p.positionQty === null) {
    tail = `Position: UNKNOWN (position read failed: ${p.positionError ?? "unknown error"}). ` +
      `Treat as open (runbook §11).`;
  } else if (p.positionQty === 0) {
    tail = `Position: flat. No immediate broker action required; investigate before resuming.`;
  } else {
    tail = `OPEN POSITION qty=${p.positionQty}: pause leaves it UNMANAGED (no re-leg, no ` +
      `flatten) until an operator acts. Use panic?action=liquidate or a manual flatten, not ` +
      `resume alone (runbook §11).`;
  }
  const message = `${head} ${tail}`;
  return {
    event_type: "equity_floor_fired",
    message,
    ticker: p.ticker,
    equity: p.equity,
    baseline: p.baseline,
    floor_pct: 0.15,
    drawdown_pct: p.drawdownPct,
    position_qty: p.positionQty,
  };
}

export function notifyEquityFloorFired(
  p: Parameters<typeof equityFloorFiredEvent>[0],
): Promise<void> {
  return notify(equityFloorFiredEvent(p));
}

export function tradeFailedEvent(p: {
  symbol: string;
  side: "BUY" | "SELL";
  qty: number;
  reason: string;
}): Record<string, unknown> {
  const message = `Trade failed: ${p.side} ${p.qty} ${p.symbol} -- ${p.reason}`;
  return {
    event_type: "trade_failed",
    message,
    symbol: p.symbol,
    side: p.side,
    qty: p.qty,
    reason: p.reason,
  };
}

export function notifyTradeFailed(p: Parameters<typeof tradeFailedEvent>[0]): Promise<void> {
  return notify(tradeFailedEvent(p));
}

export function stateDesyncEvent(p: {
  dbState: "LONG" | "CASH";
  brokerState: "LONG" | "CASH";
  symbol: string;
  actionTaken: string;
}): Record<string, unknown> {
  const message =
    `State desync on ${p.symbol}: DB=${p.dbState}, broker=${p.brokerState}. ${p.actionTaken}`;
  return {
    event_type: "state_desync",
    message,
    db_state: p.dbState,
    broker_state: p.brokerState,
    symbol: p.symbol,
    action_taken: p.actionTaken,
  };
}

export function notifyStateDesync(p: Parameters<typeof stateDesyncEvent>[0]): Promise<void> {
  return notify(stateDesyncEvent(p));
}

// Replaces notify_tws_disconnected — Alpaca is stateless REST, so this covers
// any broker/API error (connection, 5xx, auth).
export function brokerErrorEvent(
  p: { context: string; errorMsg: string },
): Record<string, unknown> {
  const message = `Broker API error (${p.context}): ${p.errorMsg}`;
  return {
    event_type: "broker_error",
    message,
    context: p.context,
    error_msg: p.errorMsg,
  };
}

export function notifyBrokerError(p: Parameters<typeof brokerErrorEvent>[0]): Promise<void> {
  return notify(brokerErrorEvent(p));
}

export function errorEvent(message: string): Record<string, unknown> {
  return { event_type: "error", message };
}

export function notifyError(message: string): Promise<void> {
  return notify(errorEvent(message));
}

export function notifyPanic(p: { action: string; result: string }): Promise<void> {
  const message = `🛑 PANIC — ${p.action}: ${p.result}`;
  return notify({ event_type: "panic", message, action: p.action, result: p.result });
}
